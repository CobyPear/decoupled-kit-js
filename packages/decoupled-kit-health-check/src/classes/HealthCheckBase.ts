import { logger } from '../utils/logger';
import {
	DecoupledMenuError,
	DecoupledRouterError,
	InvalidCMSEndpointError,
} from './errors';

export abstract class HealthCheckBase {
	abstract endpoint: string;
	abstract envVar: string;
	abstract log: typeof logger;
	/**
	 * Validate the set endpoint
	 */
	abstract validateEndpoint(): Promise<this | InvalidCMSEndpointError>;
	/**
	 * Uses `this.endpoint` to return a new URL
	 */
	abstract getURL(): URL;
	/**
	 * Check the url for a 200 response
	 * @param url - a URL with defined pathname and/or search params
	 */
	abstract checkFor200(url: URL): Promise<boolean>;
	/**
	 * Validate the menu query or endpoint
	 */
	abstract validateMenu(): Promise<this | DecoupledMenuError>;
	/**
	 * Validate the provided credentials
	 */
	abstract validateAuth(): Promise<this>;
	/**
	 * Validate the preview secret is set and preview is configured at the endpoint by fetching preview content.
	 * Should be skipped if credentials are not validated in `this.validateAuth()`.
	 */
	abstract validatePreview(): Promise<this>;
}
export abstract class DrupalHealthCheck extends HealthCheckBase {
	/**
	 * True if the Drupal instance has umami demo data
	 */
	abstract hasUmami: boolean;
	/**
	 * The clientID of the Drupal OAuth client
	 */
	abstract clientID: string | undefined;
	/**
	 * Sets `this.hasUmami` by checking the language settings endpoint
	 */
	abstract checkForUmami(): Promise<void>;
	/**
	 * Validates that the Decoupled Router module is enabled and working
	 */
	abstract validateRouter(): Promise<this | DecoupledRouterError>;
}
export abstract class WordPressHealthCheck extends HealthCheckBase {
	abstract appUsername: string | undefined;
}
