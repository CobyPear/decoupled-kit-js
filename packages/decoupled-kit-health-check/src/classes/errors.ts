class HealthCheckError extends Error {
	constructor(message: string) {
		super(`|__‚ùå ${message}`);
		// no need to show the stack for these errors
		this.stack = undefined;
	}
}

export class BackendNotSetError extends HealthCheckError {
	constructor(
		{ envVar }: { envVar: string },
		message = `No backend set: 
	The PANTHEON_CMS_ENDPOINT or ${envVar} environment variable must be set to fetch data.`,
	) {
		super(message);
	}
}

export class InvalidCMSEndpointError extends HealthCheckError {
	constructor(
		{ endpointType, endpoint }: { endpointType: string; endpoint: string },
		message = `${endpointType} could not be fetched from.
Check that ${endpoint} is valid and provides a 200 response`,
	) {
		super(message);
	}
}

export class DecoupledRouterError extends HealthCheckError {
	constructor(
		{ endpointType, endpoint }: { endpointType: string; endpoint: string },
		message = `Decoupled Router not detected for ${endpointType}.
Check that ${endpoint} is valid and provides a 200 response.
Also ensure that the Decoupled Router module is enabled.`,
	) {
		super(message);
	}
}

export class DecoupledMenuError extends HealthCheckError {
	constructor(
		{ endpointType, endpoint }: { endpointType: string; endpoint: string },
		message = `Decoupled Menu Endpoint not valid for ${endpointType}.
Check that ${endpoint} is valid and provides a 200 response.
Also ensure that the JSON:API Menu Items module is enabled.`,
	) {
		super(message);
	}
}
