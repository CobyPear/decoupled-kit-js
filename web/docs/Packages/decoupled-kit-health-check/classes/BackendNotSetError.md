---
id: "BackendNotSetError"
title: "Class: BackendNotSetError"
sidebar_label: "BackendNotSetError"
sidebar_position: 0
custom_edit_url: null
---

## Hierarchy

- `HealthCheckError`

  ↳ **`BackendNotSetError`**

## Constructors

### constructor

• **new BackendNotSetError**(`«destructured»`, `message?`)

#### Parameters

| Name | Type |
| :------ | :------ |
| `«destructured»` | `Object` |
| › `envVar` | `string` |
| `message` | `string` |

#### Overrides

HealthCheckError.constructor

#### Defined in

[packages/decoupled-kit-health-check/src/classes/errors.ts:10](https://github.com/pantheon-systems/decoupled-kit-js/blob/5049fc03/packages/decoupled-kit-health-check/src/classes/errors.ts#L10)

## Properties

### cause

• `Optional` **cause**: `unknown`

#### Inherited from

HealthCheckError.cause

#### Defined in

node_modules/.pnpm/typescript@5.2.2/node_modules/typescript/lib/lib.es2022.error.d.ts:24

___

### message

• **message**: `string`

#### Inherited from

HealthCheckError.message

#### Defined in

node_modules/.pnpm/typescript@5.2.2/node_modules/typescript/lib/lib.es5.d.ts:1068

___

### name

• **name**: `string`

#### Inherited from

HealthCheckError.name

#### Defined in

node_modules/.pnpm/typescript@5.2.2/node_modules/typescript/lib/lib.es5.d.ts:1067

___

### stack

• `Optional` **stack**: `string`

#### Inherited from

HealthCheckError.stack

#### Defined in

node_modules/.pnpm/typescript@5.2.2/node_modules/typescript/lib/lib.es5.d.ts:1069

___

### prepareStackTrace

▪ `Static` `Optional` **prepareStackTrace**: (`err`: `Error`, `stackTraces`: `CallSite`[]) => `any`

#### Type declaration

▸ (`err`, `stackTraces`): `any`

Optional override for formatting stack traces

##### Parameters

| Name | Type |
| :------ | :------ |
| `err` | `Error` |
| `stackTraces` | `CallSite`[] |

##### Returns

`any`

**`See`**

https://v8.dev/docs/stack-trace-api#customizing-stack-traces

#### Inherited from

HealthCheckError.prepareStackTrace

#### Defined in

node_modules/.pnpm/@types+node@18.16.14/node_modules/@types/node/globals.d.ts:11

___

### stackTraceLimit

▪ `Static` **stackTraceLimit**: `number`

#### Inherited from

HealthCheckError.stackTraceLimit

#### Defined in

node_modules/.pnpm/@types+node@18.16.14/node_modules/@types/node/globals.d.ts:13

## Methods

### captureStackTrace

▸ `Static` **captureStackTrace**(`targetObject`, `constructorOpt?`): `void`

Create .stack property on a target object

#### Parameters

| Name | Type |
| :------ | :------ |
| `targetObject` | `object` |
| `constructorOpt?` | `Function` |

#### Returns

`void`

#### Inherited from

HealthCheckError.captureStackTrace

#### Defined in

node_modules/.pnpm/@types+node@18.16.14/node_modules/@types/node/globals.d.ts:4
