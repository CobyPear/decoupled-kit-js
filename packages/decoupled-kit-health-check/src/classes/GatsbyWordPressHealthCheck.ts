import dotenv from 'dotenv';
import { logger } from '../utils/logger';
import { resolveDotenvFile } from '../utils/resolveDotenvFile';
import {
	BackendNotSetError,
	DecoupledMenuError,
	InvalidCMSEndpointError,
	WPGatsbyPluginError,
} from './errors';
import { WordPressHealthCheck } from './HealthCheckBase';

export class GatsbyWordPressHealthCheck extends WordPressHealthCheck {
	endpoint: string;
	envVar: string;
	appUsername: string | undefined;
	#appPassword: string | undefined;
	log = logger;
	/**
	 * Resolves .env file if it exists and check for
	 * required environment variables, and sets the endpoint
	 * @param env - process.env
	 */
	constructor({ env }: { env: typeof process.env }) {
		super();
		if (process.env.NODE_ENV !== 'production') {
			dotenv.config({
				path: resolveDotenvFile(),
			});
		} else {
			console.log(
				'Production environment detected, skipping .env* resolution.',
			);
		}
		this.appUsername = env['WP_APPLICATION_USERNAME'];
		this.#appPassword = env['WP_APPLICATION_PASSWORD'];

		console.log('Checking for PANTHEON_CMS_ENDPOINT or WPGRAPHQL_URL...');
		const keys = ['WPGRAPHQL_URL', 'PANTHEON_CMS_ENDPOINT'];
		const backendVars: {
			[key in 'WPGRAPHQL_URL' | 'PANTHEON_CMS_ENDPOINT']?: string;
		} = {};
		Object.entries(env)
			.filter(([key]) => {
				if (keys.includes(key)) {
					return true;
				}
				return false;
			})
			.forEach(([key, value]) => {
				Object.assign(backendVars, { [key]: value });
			});

		const isVarSet = Object.keys(backendVars).length > 0;
		const areMultipleVarsSet = Object.keys(backendVars).length > 1;

		if (!isVarSet) {
			throw new BackendNotSetError({ envVar: 'WPGRAPHQL_URL' });
		} else {
			const setEndpoints = Object.keys(backendVars);
			this.log.success(
				`${setEndpoints.join(' and ')} ${
					areMultipleVarsSet ? 'are' : 'is'
				} set!`,
			);
			if (areMultipleVarsSet) {
				this.log.warn(
					`Both PANTHEON_CMS_ENDPOINT and WPGRAPHQL_URL are set.\n|____Using WPGRAPHQL_URL for remaining checks.`,
				);
			}
		}

		const [[envVar, endpoint]] = areMultipleVarsSet
			? Object.entries(backendVars).filter(([key]) => key === 'WPGRAPHQL_URL')
			: Object.entries(backendVars);

		const url = /^https:\/\//.test(endpoint) ? endpoint : `https://${endpoint}`;

		this.endpoint = url;
		this.envVar = envVar;
	}
	getURL() {
		const url = new URL(this.endpoint);
		url.pathname = '/wp/graphql';
		return url;
	}
	async checkFor200(url: URL) {
		try {
			const res = await fetch(url);
			const payload = (await res.json()) as {
				errors?: unknown[];
				data?: unknown;
			};
			if (payload.errors) {
				return false;
			} else if (payload.data) {
				return true;
			}
		} catch (error) {
			void error;
		}
		return false;
	}
	async validateEndpoint() {
		console.log('Validating CMS endpoint...');
		const cmsEndpoint = this.getURL();
		cmsEndpoint.searchParams.set('query', '{__typename}');
		const endpointIsValid = await this.checkFor200(cmsEndpoint);
		if (endpointIsValid) {
			this.log.success(`${this.envVar} is valid!`);
			return this;
		}
		throw new InvalidCMSEndpointError({
			endpointType: this.envVar,
			endpoint: this.getURL().host,
		});
	}
	async validateMenu() {
		console.log('Validating Menu Item endpoint...');
		const cmsEndpoint = this.getURL();
		const footerMenuQuery = /* graphql */ `query FooterMenuQuery {
			menu(idType: NAME, id: "Example Menu") {
				menuItems {
					edges {
						node {
							id
							path
							label
						}
					}
				}
			}
		}`;
		cmsEndpoint.searchParams.set('query', footerMenuQuery);
		const menuIsValid = await this.checkFor200(cmsEndpoint);
		if (menuIsValid) {
			this.log.success('Menu Items endpoint is valid!');
			return this;
		}
		throw new DecoupledMenuError({
			endpointType: this.envVar,
			endpoint: cmsEndpoint.host,
		});
	}
	async validateAuth() {
		console.log('Validating authentication...');
		const cmsEndpoint = this.getURL();

		if (!this.appUsername || !this.#appPassword) {
			this.log.warn('Auth not valid.');
			if (!this.appUsername) {
				this.log.warn('WP_APPLICATION_USERNAME is not set.');
				this.log.suggest(
					`Get the WP_APPLICATION_USERNAME here: ðŸ”— https://${cmsEndpoint.host}/wp/wp-admin/users.php`,
				);
			}
			if (!this.#appPassword) {
				this.log.warn('WP_APPLICATION_PASSWORD is not set.');
				this.log.suggest(
					`Set a new WP_APPLICATION_PASSWORD here by clicking edit: ðŸ”— https://${cmsEndpoint.host}/wp/wp-admin/users.php`,
				);
			}
			return this;
		}
		try {
			const credentials = `${String(this.appUsername)}:${String(
				this.#appPassword,
			)}`;
			const encodedCredentials = Buffer.from(credentials, 'binary').toString(
				'base64',
			);
			const query = /* graphql */ `query LatestPostsQuery {
				posts(where: { status: PRIVATE }) {
					edges {
						node {
							id
						}
					}
				}
			}`;
			cmsEndpoint.searchParams.set('query', query);
			const res = await fetch(cmsEndpoint, {
				headers: { Authorization: `Basic ${encodedCredentials}` },
			});
			const payload = (await res.json()) as {
				data: { posts: { edges: unknown[] } };
			};
			if (payload.data.posts.edges.length > 0) {
				this.log.success('Auth is valid!');
				return this;
			}
		} catch (error) {
			void error;
		}
		this.log.warn('Auth not valid.');
		this.log.suggest(
			'Ensure the WP_APPLICATION_USERNAME and WP_APPLICATION_PASSWORD are correct.',
		);
		return this;
	}
	async validateWPGatsbyPlugin() {
		const cmsEndpoint = this.getURL();
		const query = /* graphql */ `query checkWpGatsby {
			isWpGatsby
		}`;
		cmsEndpoint.searchParams.set('query', query);
		const res = await fetch(cmsEndpoint);
		const payload = (await res.json()) as {
			data?: { isWpGatsby: boolean };
			errors?: unknown[];
		};
		if (payload.errors) {
			throw new WPGatsbyPluginError({
				envVar: this.envVar,
				endpoint: cmsEndpoint.host,
			});
		} else if (payload?.data?.isWpGatsby) {
			this.log.success('WP Gatsby plugin is active!');
			return this;
		}
		return this;
	}
	// eslint-disable-next-line @typescript-eslint/require-await
	async validatePreview() {
		return this;
	}
}
