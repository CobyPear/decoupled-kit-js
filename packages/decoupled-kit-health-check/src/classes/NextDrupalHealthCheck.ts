import dotenv from 'dotenv';
import { logger } from '../utils/logger';
import { resolveDotenvFile } from '../utils/resolveDotenvFile';
import {
	BackendNotSetError,
	DecoupledMenuError,
	DecoupledRouterError,
	InvalidCMSEndpointError,
} from './errors';
import { DrupalHealthCheck } from './HealthCheckBase';

export class NextDrupalHealthCheck extends DrupalHealthCheck {
	endpoint: string;
	envVar: string;
	clientID: string | undefined;
	#clientSecret: string | undefined;
	#previewSecret: string | undefined;
	log = logger;
	hasUmami = false;
	#access_token: string | undefined = undefined;
	/**
	 * Resolves .env file if it exists and check for
	 * required environment variables, and sets the endpoint
	 * @param env - process.env
	 */
	constructor({ env }: { env: typeof process.env }) {
		super();
		this.clientID = env['CLIENT_ID'];
		this.#clientSecret = env['CLIENT_SECRET'];
		this.#previewSecret = env['PREVIEW_SECRET'];

		console.log('Checking for PANTHEON_CMS_ENDPOINT or BACKEND_URL...');
		if (process.env.NODE_ENV !== 'production') {
			dotenv.config({
				path: resolveDotenvFile(),
			});
		} else {
			console.log(
				'Production environment detected, skipping .env* resolution.',
			);
		}
		const keys = ['BACKEND_URL', 'PANTHEON_CMS_ENDPOINT'];
		const backendVars: {
			[key in 'BACKEND_URL' | 'PANTHEON_CMS_ENDPOINT']?: string;
		} = {};
		Object.entries(env)
			.filter(([key]) => {
				if (keys.includes(key)) {
					return true;
				}
				return false;
			})
			.forEach(([key, value]) => {
				Object.assign(backendVars, { [key]: value });
			});

		const isVarSet = Object.keys(backendVars).length > 0;
		const areMultipleVarsSet = Object.keys(backendVars).length > 1;

		if (!isVarSet) {
			throw new BackendNotSetError('BACKEND_URL');
		} else {
			const setEndpoints = Object.keys(backendVars);
			this.log.success(
				`${setEndpoints.join(' and ')} ${
					areMultipleVarsSet ? 'are' : 'is'
				} set!`,
			);
			if (areMultipleVarsSet) {
				this.log.warn(
					`Both PANTHEON_CMS_ENDPOINT and BACKEND_URL are set.\n|____Using BACKEND_URL for remaining checks.`,
				);
			}
		}

		const [[envVar, endpoint]] = areMultipleVarsSet
			? Object.entries(backendVars).filter(([key]) => key === 'BACKEND_URL')
			: Object.entries(backendVars);
		this.endpoint = /^https:\/\//.test(endpoint)
			? endpoint
			: `https://${endpoint}`;
		this.envVar = envVar;
	}
	getURL() {
		return new URL(this.endpoint);
	}
	async checkFor200(url: URL) {
		try {
			const res = await fetch(url);
			if (res.status === 200) {
				return true;
			} else {
				return false;
			}
		} catch (error) {
			return false;
		}
	}
	async checkForUmami() {
		const cmsEndpoint = this.getURL();
		cmsEndpoint.pathname =
			'/jsonapi/configurable_language/configurable_language';
		this.hasUmami = await this.checkFor200(cmsEndpoint);
	}
	async validateEndpoint() {
		console.log('Validating CMS endpoint...');
		const endpointIsValid = await this.checkFor200(this.getURL());
		if (endpointIsValid) {
			this.log.success(`${this.envVar} is valid!`);
			return this;
		}
		throw new InvalidCMSEndpointError({
			endpointType: this.envVar,
			endpoint: this.getURL().host,
		});
	}
	async validateMenu() {
		console.log('Validating Menu Item endpoint...');
		const cmsEndpoint = this.getURL();
		cmsEndpoint.pathname = '/jsonapi/menu_items/footer';
		const menuIsValid = await this.checkFor200(cmsEndpoint);
		if (menuIsValid) {
			this.log.success('Menu Items endpoint is valid!');
			return this;
		}
		throw new DecoupledMenuError({
			endpointType: this.envVar,
			endpoint: this.getURL().host,
		});
	}
	async validateRouter() {
		await this.checkForUmami();
		const cmsEndpoint = this.getURL();
		cmsEndpoint.pathname = '/router/translate-path';
		cmsEndpoint.searchParams.set('format', '_json');
		this.hasUmami
			? cmsEndpoint.searchParams.set(
					'path',
					'articles/lets-hear-it-for-carrots',
			  )
			: cmsEndpoint.searchParams.set('path', 'articles/example-article');

		const isDecoupledRouterValid = await this.checkFor200(cmsEndpoint);
		if (isDecoupledRouterValid) {
			this.log.success('Decoupled Router is valid!');
			return this;
		}
		throw new DecoupledRouterError({
			endpoint: this.getURL().host,
			endpointType: this.envVar,
		});
	}

	async validateAuth() {
		console.log('Validating authentication...');
		const cmsEndpoint = this.getURL();
		if (!this.clientID) {
			this.log.warn('CLIENT_ID is required for preview but is not set.');
			this.log.suggest(
				`Get the CLIENT_ID here: üîó https://${cmsEndpoint.host}/admin/config/services/consumer`,
			);
		}
		if (!this.#clientSecret) {
			this.log.warn('CLIENT_SECRET is required for preview but is not set.');
			this.log.suggest(
				`Set a new CLIENT_SECRET here by clicking edit: üîó https://${cmsEndpoint.host}/admin/config/services/consumer`,
			);
		}
		try {
			const tokenRequestBody = {
				grant_type: 'client_credentials',
				client_id: this.clientID,
				client_secret: this.#clientSecret,
			};

			cmsEndpoint.pathname = '/oauth/token';
			Object.entries(tokenRequestBody).forEach(([key, value]) => {
				value && cmsEndpoint.searchParams.set(key, value);
			});
			const res = await fetch(cmsEndpoint, {
				method: 'POST',
				headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
				body: cmsEndpoint.searchParams,
			});
			const { access_token } = (await res.json()) as {
				access_token?: string;
			};
			if (access_token) {
				this.#access_token = access_token;
				this.log.success('Auth is valid!');
			} else {
				this.#access_token = '';
			}
		} catch (error) {
			void error;
		}
		this.log.warn('Auth not valid.');
		this.log.suggest('Ensure the CLIENT_ID and CLIENT_SECRET are correct.');

		return this;
	}
	async validatePreview() {
		if (!this.#access_token) {
			console.log(
				'‚è≠  Skipping preview endpoint validation -- authorization required.',
			);
			return this;
		}
		console.log('Checking for PREVIEW_SECRET...');
		if (!this.#previewSecret) {
			this.log.warn('PREVIEW_SECRET env var is not set.');
			this.log.suggest(
				`To set a new secret, go to üîó https://${
					this.getURL().host
				}/admin/structure/dp-preview-site and edit the preview site you want to use.`,
			);
		} else {
			this.log.success('PREVIEW_SECRET is set.');
		}

		const cmsEndpoint = this.getURL();
		try {
			cmsEndpoint.pathname = '/node/1/preview';

			const res = await fetch(cmsEndpoint, {
				headers: {
					Authorization: `Bearer ${this.#access_token}`,
				},
			});
			if (res.ok) {
				this.log.success('Preview is valid!');
			}
		} catch (error) {
			if (error instanceof Error) {
				if (
					// @ts-expect-error - cannot easily check for instanceof SystemError
					error?.cause?.code === 'ENOTFOUND' &&
					// @ts-expect-error - cannot easily check for instanceof SystemError
					typeof error?.cause?.hostname === 'string'
				) {
					this.log.warn(`Attempted to preview "${String(
						// @ts-expect-error - cannot easily check for instanceof SystemError
						error?.cause?.hostname,
					)}".
						|____ Ensure there is a preview site with that hostname configured. at üîó https://${
							cmsEndpoint.host
						}/admin/structure/dp-preview-site`);
				}
			}
		}
		this.log.warn('Could not fetch preview site.');
		return this;
	}
}
