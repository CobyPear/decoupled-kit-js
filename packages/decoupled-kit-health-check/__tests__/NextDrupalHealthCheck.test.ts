import { vi } from 'vitest';
import { NextDrupalHealthCheck } from '../src/classes/NextDrupalHealthCheck';

describe('NextDrupalHealthCheck', () => {
	beforeEach((context) => {
		context.logSpy = vi.spyOn(console, 'log');
		delete process.env['BACKEND_URL'];
		delete process.env['PANTHEON_CMS_ENDPOINT'];
		delete process.env['CLIENT_ID'];
		delete process.env['CLIENT_SECRET'];
		delete process.env['PREVIEW_SECRET'];
	});
	it('should pass for a valid BACKEND_URL and invalid auth', async ({
		logSpy,
	}) => {
		process.env['BACKEND_URL'] = 'https://drupal.test';
		const HC = new NextDrupalHealthCheck({ env: process.env });
		await HC.validateEndpoint()
			.then((hc) => hc.validateMenu())
			.then((hc) => hc.validateRouter())
			.then((hc) => hc.validateAuth())
			.then((hc) => hc.validatePreview())
			.catch(console.error);

		const result = logSpy.mock.calls.map(([call]) => call);
		expect(result).toMatchSnapshot();
	});
	it('should pass for a valid PANTHEON_CMS_ENDPOINT and invalid auth', async ({
		logSpy,
	}) => {
		process.env['PANTHEON_CMS_ENDPOINT'] = 'drupal.test';
		const HC = new NextDrupalHealthCheck({ env: process.env });
		await HC.validateEndpoint()
			.then((hc) => hc.validateMenu())
			.then((hc) => hc.validateRouter())
			.then((hc) => hc.validateAuth())
			.then((hc) => hc.validatePreview())
			.catch(console.error);

		const result = logSpy.mock.calls.map(([call]) => call);
		expect(result).toMatchSnapshot();
	});
	it('should pass for a valid backend and valid auth but no preview secret', async ({
		logSpy,
	}) => {
		process.env['PANTHEON_CMS_ENDPOINT'] = 'drupal.test';
		process.env['CLIENT_ID'] = 'abc123';
		process.env['CLIENT_SECRET'] = 'testclientsecret@1';

		const HC = new NextDrupalHealthCheck({ env: process.env });
		await HC.validateEndpoint()
			.then((hc) => hc.validateMenu())
			.then((hc) => hc.validateRouter())
			.then((hc) => hc.validateAuth())
			.then((hc) => hc.validatePreview())
			.catch(console.error);

		const result = logSpy.mock.calls.map(([call]) => call);
		expect(result).toMatchSnapshot();
	});
	it('should pass for a valid backend and valid auth', async ({ logSpy }) => {
		process.env['PANTHEON_CMS_ENDPOINT'] = 'drupal.test';
		process.env['CLIENT_ID'] = 'abc123';
		process.env['CLIENT_SECRET'] = 'testclientsecret@1';
		process.env['PREVIEW_SECRET'] = 'testpreviewsecret!';

		const HC = new NextDrupalHealthCheck({ env: process.env });
		await HC.validateEndpoint()
			.then((hc) => hc.validateMenu())
			.then((hc) => hc.validateRouter())
			.then((hc) => hc.validateAuth())
			.then((hc) => hc.validatePreview())
			.catch(console.error);

		const result = logSpy.mock.calls.map(([call]) => call);
		expect(result).toMatchSnapshot();
	});
	// Unsure why this one is considered an error
	// and only works with `it.fails` but the following has different behavior.
	it.fails(
		'should show a helpful error message if no required env vars are set',
		async () => {
			const HC = new NextDrupalHealthCheck({ env: process.env });
			await HC.validateEndpoint()
				.then((hc) => hc.validateMenu())
				.then((hc) => hc.validateRouter())
				.then((hc) => hc.validateAuth())
				.then((hc) => hc.validatePreview())
				.catch((err) => console.log(err.message));
		},
	);
	it('should show a helpful error message if the backend is invalid', async ({
		logSpy,
	}) => {
		process.env['PANTHEON_CMS_ENDPOINT'] = 'invalid.drupal.test';
		const HC = new NextDrupalHealthCheck({ env: process.env });
		await HC.validateEndpoint()
			.then((hc) => hc.validateMenu())
			.then((hc) => hc.validateRouter())
			.then((hc) => hc.validateAuth())
			.then((hc) => hc.validatePreview())
			.catch((err) => console.log(err.message));

		const result = logSpy.mock.calls.map(([call]) => call);
		expect(result).toMatchSnapshot();
	});
});
