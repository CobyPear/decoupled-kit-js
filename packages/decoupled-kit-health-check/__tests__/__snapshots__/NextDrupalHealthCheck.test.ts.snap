// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NextDrupalHealthCheck > should pass for a valid BACKEND_URL and invalid auth 1`] = `
[
  "No .env* file found, assuming production environment.",
  "Checking for PANTHEON_CMS_ENDPOINT or BACKEND_URL...",
  "‚é©‚úÖ BACKEND_URL is set!",
  "Validating CMS endpoint...",
  "‚é©‚úÖ BACKEND_URL is valid!",
  "Validating Menu Item endpoint...",
  "‚é©‚úÖ Menu Items endpoint is valid!",
  "Validating Decoupled Router module...",
  "‚é©‚úÖ Decoupled Router is valid!",
  "Validating authentication...",
  "‚é©üí° CLIENT_ID is required for preview but is not set.",
  "‚é© Get the CLIENT_ID here: üîó https://drupal.test/admin/config/services/consumer",
  "‚é©üí° CLIENT_SECRET is required for preview but is not set.",
  "‚é© Set a new CLIENT_SECRET here by clicking edit: üîó https://drupal.test/admin/config/services/consumer",
  "‚é©üí° Auth not valid.",
  "‚é© Ensure the CLIENT_ID and CLIENT_SECRET are correct.",
  "‚é©‚è≠  Skipping preview endpoint validation -- authorization required.",
]
`;

exports[`NextDrupalHealthCheck > should pass for a valid PANTHEON_CMS_ENDPOINT and invalid auth 1`] = `
[
  "No .env* file found, assuming production environment.",
  "Checking for PANTHEON_CMS_ENDPOINT or BACKEND_URL...",
  "‚é©‚úÖ PANTHEON_CMS_ENDPOINT is set!",
  "Validating CMS endpoint...",
  "‚é©‚úÖ PANTHEON_CMS_ENDPOINT is valid!",
  "Validating Menu Item endpoint...",
  "‚é©‚úÖ Menu Items endpoint is valid!",
  "Validating Decoupled Router module...",
  "‚é©‚úÖ Decoupled Router is valid!",
  "Validating authentication...",
  "‚é©üí° CLIENT_ID is required for preview but is not set.",
  "‚é© Get the CLIENT_ID here: üîó https://drupal.test/admin/config/services/consumer",
  "‚é©üí° CLIENT_SECRET is required for preview but is not set.",
  "‚é© Set a new CLIENT_SECRET here by clicking edit: üîó https://drupal.test/admin/config/services/consumer",
  "‚é©üí° Auth not valid.",
  "‚é© Ensure the CLIENT_ID and CLIENT_SECRET are correct.",
  "‚é©‚è≠  Skipping preview endpoint validation -- authorization required.",
]
`;

exports[`NextDrupalHealthCheck > should pass for a valid backend and valid auth 1`] = `
[
  "No .env* file found, assuming production environment.",
  "Checking for PANTHEON_CMS_ENDPOINT or BACKEND_URL...",
  "‚é©‚úÖ PANTHEON_CMS_ENDPOINT is set!",
  "Validating CMS endpoint...",
  "‚é©‚úÖ PANTHEON_CMS_ENDPOINT is valid!",
  "Validating Menu Item endpoint...",
  "‚é©‚úÖ Menu Items endpoint is valid!",
  "Validating Decoupled Router module...",
  "‚é©‚úÖ Decoupled Router is valid!",
  "Validating authentication...",
  "‚é©‚úÖ Auth is valid!",
  "Checking for PREVIEW_SECRET...",
  "‚é©‚úÖ PREVIEW_SECRET is set.",
  "‚é©‚úÖ Preview is valid!",
]
`;

exports[`NextDrupalHealthCheck > should pass for a valid backend and valid auth but no preview secret 1`] = `
[
  "No .env* file found, assuming production environment.",
  "Checking for PANTHEON_CMS_ENDPOINT or BACKEND_URL...",
  "‚é©‚úÖ PANTHEON_CMS_ENDPOINT is set!",
  "Validating CMS endpoint...",
  "‚é©‚úÖ PANTHEON_CMS_ENDPOINT is valid!",
  "Validating Menu Item endpoint...",
  "‚é©‚úÖ Menu Items endpoint is valid!",
  "Validating Decoupled Router module...",
  "‚é©‚úÖ Decoupled Router is valid!",
  "Validating authentication...",
  "‚é©‚úÖ Auth is valid!",
  "Checking for PREVIEW_SECRET...",
  "‚é©üí° PREVIEW_SECRET env var is not set.",
  "‚é© To set a new secret, go to üîó https://drupal.test/admin/structure/dp-preview-site and edit the preview site you want to use.",
  "‚è≠  Skipping preview endpoint validation -- PREVIEW_SECRET required.",
]
`;

exports[`NextDrupalHealthCheck > should show a helpful error message if the backend is invalid 1`] = `
[
  "No .env* file found, assuming production environment.",
  "Checking for PANTHEON_CMS_ENDPOINT or BACKEND_URL...",
  "‚é©‚úÖ PANTHEON_CMS_ENDPOINT is set!",
  "Validating CMS endpoint...",
  "‚é©‚ùå PANTHEON_CMS_ENDPOINT could not be fetched from.
Check that invalid.drupal.test is valid and provides a 200 response",
]
`;

exports[`NextDrupalHealthCheck > should show a helpful error message if the decoupled router is not available 1`] = `
[
  "No .env* file found, assuming production environment.",
  "Checking for PANTHEON_CMS_ENDPOINT or BACKEND_URL...",
  "‚é©‚úÖ PANTHEON_CMS_ENDPOINT is set!",
  "Validating CMS endpoint...",
  "‚é©‚úÖ PANTHEON_CMS_ENDPOINT is valid!",
  "Validating Menu Item endpoint...",
  "‚é©‚úÖ Menu Items endpoint is valid!",
  "Validating Decoupled Router module...",
  "‚é©‚ùå Decoupled Router not detected for PANTHEON_CMS_ENDPOINT.
Check that invalid.drupal.test is valid and provides a 200 response.
Also ensure that the Decoupled Router module is enabled.",
]
`;
